documentation/man/hmmindex.man
.TH "hmmindex" 1 @RELEASEDATE@ "HMMER @RELEASE@" "HMMER Manual"

.SH NAME
.TP 
hmmindex - create a binary SSI index for an HMM database

.SH SYNOPSIS
.B hmmindex
.I [options]
.I database

.SH DESCRIPTION

.B hmmindex
is a utility that creates a binary SSI ("squid sequence index"
format) index for an HMM database file called
.I database.
The new index file is named
.IR database.ssi.
An SSI index file is required for 
.B hmmfetch
to work, and also for the PVM implementation of 
.B hmmpfam.

.SH OPTIONS

.TP
.B -h
Print brief help; includes version number and summary of
all options, including expert options.


.SH SEE ALSO

.PP
Master man page, with full list of and guide to the individual man
pages: see 
.B hmmer(1).
.PP
A User guide and tutorial came with the distribution:
.B Userguide.ps
[Postscript] and/or
.B Userguide.pdf
[PDF].
.PP
Finally, all documentation is also available online via WWW: 
.B http://hmmer.wustl.edu/

.SH AUTHOR

This software and documentation is: 
.nf
@COPYRIGHT@
@LICENSE@
.fi
See the file COPYING in your distribution for complete details.

.nf
Sean Eddy
HHMI/Dept. of Genetics
Washington Univ. School of Medicine
4566 Scott Ave.
St Louis, MO 63110 USA
Phone: 1-314-362-7666
FAX  : 1-314-362-7855
Email: eddy@genetics.wustl.edu
.fi


release-notes/HISTORY

# HMMER1, MRC-LMB Cambridge UK and WashU Genetics, Saint Louis
#
1.5         29 Apr 1994   Local release in Cambridge    
1.6         28 Jul 1994   First public release          
1.7         27 Sep 1994                                 
1.7.4       10 Nov 1994                                 
1.8         17 Apr 1995   First release from US
1.8.3       23 Jun 1997                                 
1.8.4       25 Jun 1997                                 


# HMMER2 development, WashU Genetics, Saint Louis
#
2.0c-alpha   3 Feb 1998
2.0e-alpha  11 Mar 1998
2.0 beta    28 May 1998



# HMMER2 stable, WashU Genetics, Saint Louis
#
2.1         14 Oct 1998  First public release of 2.x series
2.1.1       11 Dec 1998
2.1.2        2 May 1999
2.1.3        4 Dec 1999
2.2          4 May 2001
2.3         23 Apr 2003
2.3.1       12 Jun 2003
2.3.2        3 Oct 2003
2.3.3        2 Aug 2008


## HMMER3 development, HHMI Janelia Farm
#
3.0a1       12 Jan 2009
3.0a2	    12 Mar 2009
3.0b1	    17 Jun 2009
3.0b2 	    21 Jun 2009
3.0b3	    13 Nov 2009
3.0rc1	    10 Feb 2010
3.0rc2	     8 Mar 2010


## HMMER3 stable, HHMI Janelia Farm and Harvard University
#
3.0            Mar 2010
3.1b1       22 Apr 2013  (Travis Wheeler, buildmaster)
3.1b2	     5 Mar 2015  (Travis Wheeler, buildmaster)






src/hmmindex.c
/* hmmindex.c
 * Create an SSI index file for an HMM database.
 * 
 * SRE, Wed Aug  5 11:05:03 1998 [St. Louis]
 * SVN $Id: hmmindex.c 1382 2005-05-03 22:25:36Z eddy $
 */

#include "config.h"
#include "squidconf.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "squid.h"
#include "ssi.h"

#include "plan7.h"
#include "structs.h"
#include "funcs.h"
#include "globals.h"


static char banner[] = "hmmindex -- create SSI index for an HMM database";

static char usage[] = "\
Usage: hmmindex [-options] <hmmfile>\n\
Available options are:\n\
  -h             : print short usage and version info, then exit\n\
";

static char experts[] = "\
";

static struct opt_s OPTIONS[] = {
   { "-h", TRUE, sqdARG_NONE  },
};
#define NOPTIONS (sizeof(OPTIONS) / sizeof(struct opt_s))

int
main(int argc, char **argv)
{
  char    *hmmfile;             /* HMM file to open                */
  SSIINDEX *ssi;                /* SSI index in memory             */
  char    *ssifile;             /* name of SSI index on disk       */
  HMMFILE *hmmfp;               /* opened hmm file pointer         */
  struct plan7_s     *hmm;      /* a hidden Markov model           */
  int     nhmm;		        /* counter over HMMs               */
  int     npri, nsec;		/* # of names, accessions          */
  int     fh;			/* file handle                     */
  int     status;		/* return status from SSI call     */

  char *optname;		/* name of option found by Getopt() */
  char *optarg;			/* argument found by Getopt()       */
  int   optind;		        /* index in argv[]                  */

  /***********************************************
   * Parse the command line
   ***********************************************/

  while (Getopt(argc, argv, OPTIONS, NOPTIONS, usage,
		&optind, &optname, &optarg))
    {
      if (strcmp(optname, "-h")        == 0)
	{
	  HMMERBanner(stdout, banner);
	  puts(usage);
	  puts(experts);
	  exit(0);
	}
    }

  if (argc - optind != 1) Die("Incorrect number of arguments.\n%s\n", usage);
  hmmfile = argv[optind++];

  /***********************************************
   * Open our input HMM file, make sure all is well with the output SSI filename
   ***********************************************/

  if ((hmmfp = HMMFileOpen(hmmfile, NULL)) == NULL)
    Die("failed to open HMM file %s for reading.", hmmfile);
  if (hmmfp->ssi != NULL)
    Die("SSI index already exists for %s.\nPlease delete it first.", hmmfile);
  
  ssifile = MallocOrDie(strlen(hmmfile) + 5);
  sprintf(ssifile, "%s%s", hmmfile, ".ssi");
  if (FileExists(ssifile))   /* shouldn't happen */
    Die("An SSI file %s already exists; please delete it first", ssifile);

  if ((ssi = SSICreateIndex(hmmfp->mode)) == NULL)
    Die("Failed to initialize the SSI index structure");
  if (SSIAddFileToIndex(ssi, hmmfile, hmmfp->is_binary, &fh) != 0)
    Die("SSIAddFileToIndex() failed");

  /*********************************************** 
   * Show the banner
   ***********************************************/

  HMMERBanner(stdout, banner);
  printf("HMM file:                 %s\n", hmmfile);
  if (hmmfp->mode == SSI_OFFSET_I64) 
    printf("Index file mode:          64-bit (large HMM file)\n");
  printf("- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n\n");

  /***********************************************
   * Get offsets and names for every model; store in keylist
   ***********************************************/

  printf("Determining offsets for %s, please be patient...\n", hmmfile);

  nhmm = npri = nsec = 0;
  while (HMMFileRead(hmmfp, &hmm)) 
    {	
      if (hmm == NULL) 
	Die("HMM file %s may be corrupt or in incorrect format; parse failed", hmmfile);

				/* record name of HMM as the primary retrieval key */
      status = SSIAddPrimaryKeyToIndex(ssi, hmm->name, fh, &(hmmfp->offset), NULL, 0);
      if (status != 0) Die("SSIAddPrimaryKeyToIndex() failed");
      npri++;

				/* record accession of HMM as a secondary retrieval key */
      if (hmm->flags & PLAN7_ACC) {
	status = SSIAddSecondaryKeyToIndex(ssi, hmm->acc, hmm->name);
	if (status != 0) Die("SSIAddSecondaryKeyToIndex() failed");
	nsec++;
      }

      nhmm++;
      FreePlan7(hmm);
    }
  HMMFileClose(hmmfp);

  /***********************************************
   * Output the SSI file
   ***********************************************/

  status = SSIWriteIndex(ssifile, ssi);
  if (status != 0) Die("SSIWriteIndex() failed");

  printf("Complete.\n");
  printf("HMM file:       %s\n", hmmfile);
  printf("SSI index:      %s\n", ssifile);
  printf("# of HMMS:      %d\n", nhmm);
  printf("HMM names:      %d\n", npri);
  printf("HMM accessions: %d\n", nsec);


  /***********************************************
   * Exit
   ***********************************************/

  free(ssifile);
  SSIFreeIndex(ssi);
  SqdClean();
  return 0;
}


/************************************************************
 * @LICENSE@
 ************************************************************/

